tutorials/haskell/gui-application/tutorial.md

```diff
@@ -0,0 +1,825 @@
+  mkBtn "MC"  >>= attach 0 1 1 1   -- (5)
+  mkBtn "MR"  >>= attach 1 1 1 1
+  mkBtn "MS"  >>= attach 2 1 1 1
+  mkBtn "M+"  >>= attach 3 1 1 1
+  mkBtn "M–"  >>= attach 4 1 1 1
+  mkBtn "←"   >>= attach 0 2 1 1
+  mkBtn "CE"  >>= attach 1 2 1 1
+  mkBtn "C"   >>= attach 2 2 1 1
+  mkBtn "±"   >>= attach 3 2 1 1
+  mkBtn "√"   >>= attach 4 2 1 1
+  mkBtn "7"   >>= attach 0 3 1 1
+  mkBtn "8"   >>= attach 1 3 1 1
+  mkBtn "9"   >>= attach 2 3 1 1
+  mkBtn "÷"   >>= attach 3 3 1 1
+  mkBtn "%"   >>= attach 4 3 1 1
+  mkBtn "4"   >>= attach 0 4 1 1
+  mkBtn "5"   >>= attach 1 4 1 1
+  mkBtn "6"   >>= attach 2 4 1 1
+  mkBtn "*"   >>= attach 3 4 1 1
+  mkBtn "1/x" >>= attach 4 4 1 1
+  mkBtn "1"   >>= attach 0 5 1 1
+  mkBtn "2"   >>= attach 1 5 1 1
+  mkBtn "3"   >>= attach 2 5 1 1
+  mkBtn "–"   >>= attach 3 5 1 1
+  mkBtn "="   >>= attach 4 5 1 2
+  mkBtn "0"   >>= attach 0 6 2 1
+  mkBtn "."   >>= attach 2 6 1 1
+  mkBtn "+"   >>= attach 3 6 1 1
+  containerAdd window grid         -- (6)
+  …
+```
+
+1. `gridNew` creates a new grid.
+
+2. `gridSetRowHomogeneous grid True` makes every row have equal height.
+
+3. Here we define the `attach` helper function. It attaches given widget to
+   our `grid`. The argument order of this function helps to use it with
+   `(>>=)`.
+
+4. We attach `display` we created previously to the `grid`. It will occupy
+   the entire top row.
+
+5. Here we use combination of `mkBtn` helper and `attach` to quickly create
+   buttons and place them on the grid. I'll show `mkBtn` is a moment.
+
+6. Now the grid itself needs to be inserted into `window` to be visible.
+   This is done with help of above-mentioned `containerAdd` function.
+
+`mkBtn` is a helper for button creation, right now it's very simple:
+
+```haskell
+mkBtn :: String -> IO Button
+mkBtn label = do
+  btn <- buttonNew
+  set btn [ buttonLabel := label ]
+  return btn
+```
+
+We create new button, set its attributes (just label in our case) and return
+the button.
+
+![Our calculator looks like a real calculator](/tutorials/haskell/gui-application/calc-2.png)
+
+## Signals and events
+
+The application looks like a calculator but does not behave like one yet. To
+fix this, we need to learn about *signals* and *events*.
+
+*Signal* is a name for things that may happen on the form. Almost always
+signals are connected with user's actions. An example of signal is *focus* —
+the moment when a widget becomes active on the form.
+
+To execute some code on signal we use the `on` helper:
+
+```haskell
+on :: object -> Signal object callback -> callback -> IO (ConnectId object)
+```
+
+Here `object` is the widget of interest, `callback` is the action that we
+want to perform. The `on` function returns `ConnectId` parametrized over the
+`object` type (so we cannot mix up connection identifiers for different
+types of objects). This is the identifier of signal handler and its sole
+purpose to give you a way to disconnect a signal handler if you ever need
```

Comment:
@mrkkrp "its sole purpose **is** to give you". Below, "`disconnect` from" or "the `disconnect` **function** from".